File: Coding_Test_with_Python/4. 구현/4-4.py
raw, col = map(int, input().split())

count = 1
x, y, direction = list(map(int, input().split())) # 행, 열, 방향
# 북 동 남 서
dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]

game_map = [list(map(int, input().split())) for _ in range(raw)]
game_map[x][y] = 2

def turn_left():
    global direction
    direction -= 1
    if direction == -1:
        direction = 3

turn_time = 0
while True:
    turn_left()
    next_x = x + dx[direction]
    next_y = y + dy[direction]
    # 회전한 이후 정면에 가보지 않은 칸이 존재하는 경우 이동
    if game_map[next_x][next_y] == 0:
        game_map[next_x][next_y] = 2
        x = next_x
        y = next_y
        count += 1
        turn_time = 0
        continue
    # 회전 이후 정면에 가보지 않은 칸이 없거나 바다인 경우
    else:
        turn_time += 1
    # 네 곳 다 막힌 경우
    if turn_time == 4:
        next_y = y - dy[direction]
        next_x = x - dx[direction]
        # 뒤로 갈 수 있는 경우
        if game_map[next_x][next_y] == 2:
            x = next_x
            y = next_y
        else:
            break
        turn_time = 0


print(count)

File: Coding_Test_with_Python/4. 구현/4-4.cpp
#include <iostream>
using namespace std;

int N, M; // row col
int location_x, location_y, direction;

int direction_x[4] = {0, 1, 0, -1};
int direction_y[4] = {-1, 0, 1, 0};

int **map;
int answer = 1;

void setMap(int row, int col) {
    map = new int*[row];
    for (int i = 0; i < row; i++) {
        map[i] = new int[col];
    }
}

void print() {
    for (int i = 0 ; i < N; i++) {
        for (int j = 0; j < M; j++) {
            cout << map[i][j] << " ";
        }
        cout << endl;
    }
}


void setGoneLocation(int y, int x) {
    map[y][x] = 2;
    print();
}

void input() {
    // 맵 생성 
    cin >> N >> M;
    setMap(N, M);

    cin >> location_x >> location_y >> direction;

    for (int i = 0; i < N; i++) {
        for (int j = 0; j < M; j++) {
            cin >> map[i][j];
        }
    }

    setGoneLocation(location_y, location_x);

}

bool checkMapRange(int x, int y) {
    if (x < 0 || x >= M || y < 0 || y >= N) {
        return false;
    }
    return true;
}

void setDirection (int &direction) {
    direction--;
    if (direction == -1) {
        direction = direction+4;
    }
}

bool canMove() {
    int count = 4;
    while(count != 0) {
        
        int target_location_x = location_x + direction_x[direction];
        int target_location_y = location_y + direction_y[direction];

        if (checkMapRange(target_location_x, target_location_y) && map[target_location_y][target_location_x] == 0) {
            // todo : functional 
            location_x = target_location_x;
            location_y = target_location_y;
            setGoneLocation(location_y, location_x);
            return true;
        }        

        setDirection(direction);
        count--;
    }
    return false;
}

void solve() {
    while (canMove()) {
        answer++;
    }
}

int main() {

    input();
    
    solve();
    cout << answer << endl;

    return 0;
}